core/shared/src/main/scala/fs2/util/Concurrent.scala

```diff
@@ -0,0 +1,299 @@
+                ec.executeThunk { cb((r: Either[Throwable, A]).map((_,id2))) }
+              }
+              waiting = LinkedMap.empty
+              result = r
+              cb(Right(true))
+            }
+            else cb(Right(false))
+
+          case Msg.Nevermind(id, cb) =>
+            val interrupted = waiting.get(id).isDefined
+            waiting = waiting - id
+            ec.executeThunk { cb(Right(interrupted)) }
+        }
+
+        new Ref[F, A] {
+          implicit val F: Concurrent[F] = self
+
+          def access: F[(A, Attempt[A] => F[Boolean])] =
+            F.delay(new MsgId).flatMap { mid =>
+              getStamped(mid).map { case (a, id) =>
+                val set = (a: Attempt[A]) =>
+                  F.async[Boolean] { cb => actor ! Msg.TrySet(id, a, cb) }
+                (a, set)
+              }
+            }
+
+          /**
+           * Return a `F` that submits `t` to this ref for evaluation.
+           * When it completes it overwrites any previously `put` value.
+           */
+          def set(t: F[A]): F[Unit] =
+            F.liftIO(F.runAsync(F.shift(t)(ec)) { r => IO(actor ! Msg.Set(r)) })
+
+          private def getStamped(msg: MsgId): F[(A,Long)] =
+            F.async[(A,Long)] { cb => actor ! Msg.Read(cb, msg) }
+
+          /** Return the most recently completed `set`, or block until a `set` value is available. */
+          override def get: F[A] = F.delay(new MsgId).flatMap { mid => getStamped(mid).map(_._1) }
+
+          /** Like `get`, but returns a `F[Unit]` that can be used cancel the subscription. */
+          def cancellableGet: F[(F[A], F[Unit])] = F.delay {
+            val id = new MsgId
+            val get = getStamped(id).map(_._1)
+            val cancel = F.async[Unit] {
+              cb => actor ! Msg.Nevermind(id, r => cb((r: Either[Throwable, Boolean]).map(_ => ())))
+            }
+            (get, cancel)
+          }
+
+          /**
+           * Runs `f1` and `f2` simultaneously, but only the winner gets to
+           * `set` to this `ref`. The loser continues running but its reference
+           * to this ref is severed, allowing this ref to be garbage collected
+           * if it is no longer referenced by anyone other than the loser.
+           */
+          def race(f1: F[A], f2: F[A]): F[Unit] = F.delay {
+            val ref = new AtomicReference(actor)
+            val won = new AtomicBoolean(false)
+            val win = (res: Attempt[A]) => {
+              // important for GC: we don't reference this ref
+              // or the actor directly, and the winner destroys any
+              // references behind it!
+              if (won.compareAndSet(false, true)) {
+                val actor = ref.get
+                ref.set(null)
+                actor ! Msg.Set(res)
+              }
+            }
+            unsafeRunAsync(f1)(res => IO(win(res)))
+            unsafeRunAsync(f2)(res => IO(win(res)))
```

Comment:
If either `f1` or `f2` is fully synchronous, then this function will not run them in parallel!  I think you either need to inject a `shift`, or make it clear in the documentation that the effects *may* be run simultaneously.